RATIS-1457. DataStreamManagement startTransaction error handler

guohao-rosicky opened the pull request:
## What changes were proposed in this pull request?

DataStreamManagement  startTransaction error handler, doing so will fetch exceptions that may occur in async

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/RATIS-1457

szetszwo commented on ratis-netty/src/main/java/org/apache/ratis/netty/server/DataStreamManagement.java:
```diff
@@ -323,16 +323,16 @@ static void sendReply(List<CompletableFuture<DataStreamReply>> remoteWrites,
   }
 
-  private CompletableFuture<Void> startTransaction(StreamInfo info, DataStreamRequestByteBuf request, long bytesWritten,
+  private void startTransaction(StreamInfo info, DataStreamRequestByteBuf request, long bytesWritten,
```
Let's keep it async and sync it outside.
```
+++ b/ratis-netty/src/main/java/org/apache/ratis/netty/server/DataStreamManagement.java
@@ -409,7 +409,7 @@ public class DataStreamManagement {
           } else if (close) {
             if (info.isPrimary()) {
               // after all server close stream, primary server start transaction
-              startTransaction(info, request, bytesWritten, ctx);
+              startTransaction(info, request, bytesWritten, ctx).join();
             } else {
               sendReply(remoteWrites, request, bytesWritten, info.getCommitInfos(), ctx);
             }
```

guohao-rosicky replied:
Has been modified

szetszwo reviewed (commented):
@guohao-rosicky , thanks for working on this.  Just a minor comment.

szetszwo commented:
The test failure seems related.  Please take a look.

guohao-rosicky commented:
Seems to have nothing to do with the submitted code
```
Errors: 
Error:    TestFileStoreStreamingWithGrpcCluster>FileStoreStreamingBaseTest.testFileStoreStreamMultipleFiles:100->FileStoreStreamingBaseTest.testMultipleFiles:138 » TestTimedOut
Error:    TestFileStoreStreamingWithGrpcCluster>FileStoreStreamingBaseTest.testFileStoreStreamSingleFile:78->FileStoreStreamingBaseTest.testSingleFile:116 » TestTimedOut
```

szetszwo commented:
@guohao-rosicky , try revert the change.  These tests will pass immediately.

guohao-rosicky commented:
I reworked the code to get exceptions without having to wait sync, which was low performance. Added whenComplate method to handle exceptions and send exception messages

@szetszwo Please take a look.

szetszwo reviewed (approved):
+1 the change looks good.
